subsys/nrf_security/src/ssf_secdom/sdfw_psa_ns_api.c

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2025 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdint.h>
+#include <stdbool.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/cache.h>
+
+#include <psa/client.h>
+#include <psa/error.h>
+
+#include <tfm/sdfw_psa_ipc_service.h>
+
+#include "sdfw_psa_ns_ipc.h"
+#include "bounce_buffers.h"
+
+/* The correctness of the serialization depends on these asserts */
+BUILD_ASSERT(4 == sizeof(psa_handle_t));
+BUILD_ASSERT(4 == sizeof(int32_t));
+BUILD_ASSERT(4 == sizeof(const psa_invec *));
+BUILD_ASSERT(4 == sizeof(size_t));
+BUILD_ASSERT(4 == sizeof(psa_outvec *));
+BUILD_ASSERT(4 == sizeof(psa_status_t *));
+
+psa_status_t psa_call_buffered_and_flushed(psa_handle_t handle, int32_t type,
+					   const psa_invec *in_vec, size_t in_len,
+					   psa_outvec *out_vec, size_t out_len)
+{
+	/* We have no need for this at this time */
+	ARG_UNUSED(type);
+
+	psa_status_t ipc_status = sdfw_psa_ns_ipc_setup();
+
+	if (ipc_status != PSA_SUCCESS) {
+		return ipc_status;
+	}
+
+	/* volatile and flushed because the cpusec core will usually
+	 * modify this variable
+	 */
+	psa_status_t volatile status = PSA_ERROR_COMMUNICATION_FAILURE;
+	sys_cache_data_flush_range((void *)&status, sizeof(status));
+
+	uint32_t ipc_service_buf[SDFW_PSA_IPC_DATA_LEN];
+
+	ipc_service_buf[0] = handle; /* i.e. TFM_CRYPTO_HANDLE defined to 0x40000100U */
+	ipc_service_buf[1] = (uint32_t)in_vec;
+	ipc_service_buf[2] = in_len;
+	ipc_service_buf[3] = (uint32_t)out_vec;
+	ipc_service_buf[4] = out_len;
+	ipc_service_buf[5] = (uint32_t)&status;
+
+	int32_t ret = sdfw_psa_ns_ipc_send(ipc_service_buf, sizeof(ipc_service_buf));
+
+	if (ret != sizeof(ipc_service_buf)) {
+		return PSA_ERROR_COMMUNICATION_FAILURE;
+	}
+
+	do {
+		sys_cache_data_flush_and_invd_range((void *)&status, sizeof(status));
+	} while (status == PSA_ERROR_COMMUNICATION_FAILURE);
```

Comment:
I hope the new ipc layer will provide a more robust waiting mechanism for a blocking call, can you maybe leave a comment about this being removed once there is a blocking call implemented?

Reply:
I'm pretty sure this will be completely rewritten when we get the new ipc layer so there is no risk and no need for a comment IMHO.